pkg/controller/registrysyncer/registrysyncer.go

```diff
@@ -0,0 +1,316 @@
+			continue
+		}
+		client := r.registryClients[ctx]
+		if err := client.Get(r.ctx, types.NamespacedName{Name: req.Namespace}, &corev1.Namespace{}); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return fmt.Errorf("failed to check if namespace %s exists: %w", req.Namespace, err)
+			}
+			if err := client.Create(r.ctx, &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: req.Namespace}}); err != nil && !apierrors.IsAlreadyExists(err) {
+				return fmt.Errorf("failed to create namespace %s: %w", req.Namespace, err)
+			}
+		}
+
+		if err := r.ensureImageStream(sourceImageStream, client, log); err != nil {
+			return fmt.Errorf("failed to ensure imagestream: %w", err)
+		}
+
+		if isTag != nil && isTag.Image.Name == sourceImageStreamTag.Image.Name {
+			log.Debug("ImageStreamTag is current")
+			return nil
+		}
+
+		if err := r.ensureImagePullSecret(req.Namespace, client, log); err != nil {
+			return fmt.Errorf("failed to ensure imagePullSecret: %w", err)
+		}
+		imageName, err := publicDomainForImage(srcContext, sourceImageStreamTag.Image.DockerImageReference)
+		if err != nil {
+			return fmt.Errorf("failed to get the public domain: %w", err)
+		}
+
+		imageStreamImport := &imagev1.ImageStreamImport{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: req.Namespace,
+				Name:      imageStreamName,
+			},
+			Spec: imagev1.ImageStreamImportSpec{
+				Import: true,
+				Images: []imagev1.ImageImportSpec{{
+					From: corev1.ObjectReference{
+						Kind: "DockerImage",
+						Name: imageName,
+					},
+					To: &corev1.LocalObjectReference{Name: imageTag},
+					ReferencePolicy: imagev1.TagReferencePolicy{
+						Type: imagev1.LocalTagReferencePolicy,
+					},
+				}},
+			},
+		}
+
+		r.log.Debugf("creating import %v on cluster %s", imageStreamImport, ctx)
+
+		// ImageStreamImport is not an ordinary api but a virtual one that does the import synchronously
+		if err := client.Create(r.ctx, imageStreamImport); err != nil {
+			return fmt.Errorf("failed to import Image: %w", err)
+		}
+
+		// This should never be needed, but we shouldn't panic if the server screws up
+		if imageStreamImport.Status.Images == nil {
+			imageStreamImport.Status.Images = []imagev1.ImageImportStatus{{}}
+		}
+		if imageStreamImport.Status.Images[0].Image == nil {
+			return fmt.Errorf("imageStreamImport did not succeed: reason: %s, message: %s", imageStreamImport.Status.Images[0].Status.Reason, imageStreamImport.Status.Images[0].Status.Message)
+		}
+
+		log.Debug("Imported successfully")
+	}
+	return nil
+}
+
+func findNewest(isTags map[string]*imagev1.ImageStreamTag) (string, error) {
+	resultContext := ""
+	t := metav1.NewTime(time.Time{})
+	for ctx, isTag := range isTags {
+		if isTag == nil {
+			continue
+		}
+		if t.Before(&isTag.Image.CreationTimestamp) {
+			t = isTag.Image.CreationTimestamp
+			resultContext = ctx
+		}
+	}
+	if resultContext != "" {
+		return resultContext, nil
+	}
+	return "", fmt.Errorf("cannot find the newest isTag from %v", isTags)
```

Comment:
Thats not an error but expected to happen occasionally

Reply:
Updated.